src/crypto/nss.rs

```diff
@@ -87,123 +96,445 @@ pub(crate) fn serialize_key(_curve: ECDSACurve, key: &[u8]) -> Result<(ByteBuf,
+            ecdh_x_coord_bytes.len() as i32,
+        )
+    };
+    Ok(shared_secret.to_vec())
 }
 
 /// Encrypts a plaintext to produce a ciphertext, which may be longer than the plaintext.
 /// The plaintext is restricted to being a multiple of the AES block size (16 bytes) in length.
-pub(crate) fn encrypt(
-    key: &[u8],
-    plain_text: &[u8], /*PlainText*/
-) -> Result<Vec<u8> /*CypherText*/> {
-    crypt_helper(key, plain_text, Operation::Encrypt)
+pub(crate) fn encrypt(key: &[u8], data: &[u8]) -> Result<Vec<u8>> {
+    nss_gk_api::init();
+
+    if key.len() != 32 {
+        return Err(BackendError::NSSError(
+            "Invalid AES-256 key length".to_string(),
+        ));
+    }
+
+    // The input must be a multiple of the AES block size, 16
+    if data.len() % (AES_BLOCK_SIZE as usize) != 0 {
+        return Err(BackendError::NSSError(
+            "Input to encrypt is too long".to_string(),
+        ));
+    }
+    let in_len = c_uint::try_from(data.len())?;
+
+    let slot = Slot::internal()?;
+
+    let sym_key = unsafe {
+        PK11_ImportSymKey(
+            *slot,
+            CKM_AES_CBC,
+            PK11Origin::PK11_OriginUnwrap,
+            CKA_ENCRYPT,
+            SECItemBorrowed::wrap(key).as_mut(),
+            ptr::null_mut(),
+        )
+        .into_result()?
+    };
+
+    let iv = [0u8; AES_BLOCK_SIZE as usize];
+    let mut params = SECItemBorrowed::wrap(&iv);
+    let params_ptr: *mut SECItem = params.as_mut();
+    let mut out_len: c_uint = 0;
+    let mut out = vec![0; in_len as usize];
+    unsafe {
+        PK11_Encrypt(
+            *sym_key,
+            CKM_AES_CBC,
+            params_ptr,
+            out.as_mut_ptr(),
+            &mut out_len,
+            in_len,
+            data.as_ptr(),
+            in_len,
+        )
+        .into_result()?
+    }
+    // CKM_AES_CBC should have output length equal to input length.
+    debug_assert_eq!(out_len, in_len);
+
+    Ok(out)
 }
 
 /// Decrypts a ciphertext and returns the plaintext.
-pub(crate) fn decrypt(
-    key: &[u8],
-    cypher_text: &[u8], /*CypherText*/
-) -> Result<Vec<u8> /*PlainText*/> {
-    crypt_helper(key, cypher_text, Operation::Decrypt)
-}
+pub(crate) fn decrypt(key: &[u8], data: &[u8]) -> Result<Vec<u8>> {
+    nss_gk_api::init();
+    let slot = Slot::internal()?;
 
-fn crypt_helper(key: &[u8], input: &[u8], operation: Operation) -> Result<Vec<u8>> {
-    // Spec says explicitly IV=0
-    let iv = [0u8; 16];
-
-    // TODO(MS): aes_cbc_encrypt does padding, which we don't want. So we have to use common_encrypt
-    // let plain_text = aes_cbc_crypt(key.key(), &iv, cypher_text, Operation::Decrypt)
-    //     .map_err(|e| rc_crypto::Error::from(e))?;
-    let mut params = nss_sys::SECItem {
-        type_: nss_sys::SECItemType::siBuffer as u32,
-        data: iv.as_ptr() as *mut c_uchar,
-        len: c_uint::try_from(iv.len()).map_err(|_| BackendError::TryFromError)?,
+    if key.len() != 32 {
```

Comment:
Do we not have an AES key length constant?

Reply:
There's `AES_256_KEY_LENGTH` in blapi. But it's not exposed in nss-gk-api.